### PR TITLE
[DS-3650] Add subcommunities link to community object

### PR DESCRIPTION
**Pull request by terrywbrady**
https://jira.duraspace.org/browse/DS-3650

### CODE REVIEW

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/CommunityConverter.java [hunk 1]
```diff
@@ -32,6 +33,9 @@
     @Autowired
     private CollectionConverter collectionConverter;
 
+    @Autowired
+    private CommunityConverter communityConverter;
```
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/CommunityConverter.java]: this is not really needed as the current object is THE community converter. 
Please note that It must be a singleton as it is annotated with component

**Review by abollini — CHANGES_REQUESTED** [review]
Hi @terrywbrady, thanks for the PR 
Can I ask you to provide Integration test to check that new properties are exposed under the /api/communities/:uuid/subCommunities path and embedded in the response for a parent community /api/communities/:uuid?
The initialization code in org.dspace.app.rest.CommunityRestRepositoryIT.findAllTest() should be the same that you need in your new test methods

**Comment by tdonohue** [comment]
This is looking good (pending integration tests)!

I'll also note that it'd be good to add a corresponding PR to our REST Contract to describe this change (as best we can, we attempt to keep the REST Contract & code "in sync"). It'd likely just involve minor updates to the `communities` endpoint documentation: https://github.com/DSpace/Rest7Contract/blob/master/communities.md

**Comment by terrywbrady** [comment]
@abollini , for the subcommunities code, have I initialized the objects as you expect?  As I think about this, will the instantiation of a top level community trigger the instantiation of all of the child communities (and collections)?  Should I have created a link to the contained communities rather than embed them in the object?

**Comment by abollini** [comment]
IMHO the current @terrywbrady implementation is right and we will end embedding the full hierarchy in the response. This will be revisited for sure later when we will introduce the support for projections, see past wip:
https://github.com/DSpace/DSpace/pull/1688
https://github.com/DSpace/DSpace/pull/1847
where the client can ask to include or exclude specific relations in the response (and maybe limit the depth).
it is risky and not useful try to apply a workaround now. The angular-ui will be confused to get (and caching) a subcommunity json without the 2nd level children. 
A first next step could be address the issue https://jira.duraspace.org/browse/DS-3898 (it is about bitstreams in the item but obviously is the same than collections or subCommunities in a community)

**Comment by terrywbrady** [comment]
@abollini , this is available for review.  Is there any value in having an additional use case that matches the results of /api/communities/:uuid/subcommunities with /api/communities/search/subCommunities?parent=:uuid?

File: dspace-spring-rest/src/test/java/org/dspace/app/rest/CommunityRestRepositoryIT.java [hunk 2]
```diff
@@ -187,12 +190,20 @@ public void findOneRelsTest() throws Exception {
...
         getClient().perform(get("/api/core/communities/" + child1.getID().toString() + "/collections"))
                    .andExpect(status().isOk())
                    .andExpect(content().contentType(contentType));
+
+        //child1 subcommunity has no subcommunities, therefore contentType is not set
+        getClient().perform(get("/api/core/communities/" + parentCommunity.getID().toString() + "/subcommunities"))
+                   .andExpect(status().isOk());
+
+        getClient().perform(get("/api/core/communities/" + child1.getID().toString() + "/subcommunities"))
+                   .andExpect(status().isOk());
```
Comment (abollini) [inline dspace-spring-rest/src/test/java/org/dspace/app/rest/CommunityRestRepositoryIT.java]: This should return a 204 but currently it returns a wrong 200, see also https://jira.duraspace.org/browse/DS-3808

I suggest to prepare the test for 204 and keep it commented with a note //FIXME enable when DS-3808 is solved

**Review by abollini — CHANGES_REQUESTED** [review]
Only the change to the test is required after that we can merge this first implementation and (maybe revisit it later with the LinkRestRepository approach)

**Review by abollini — APPROVED** [review]
the code looks good by inspection. I will give it a try tomorrow and merge if not done by other .

**Pull request merged**